OpenRA.Mods.Common/Traits/Buildings/Building.cs

```diff
@@ -58,9 +68,97 @@ public class BuildingInfo : ITraitInfo, IOccupySpaceInfo, IPlaceBuildingDecorati
 
 		public virtual object Create(ActorInitializer init) { return new Building(init, this); }
 
+		protected static object LoadFootprint(MiniYaml yaml)
+		{
+			var footprintYaml = yaml.Nodes.FirstOrDefault(n => n.Key == "Footprint");
+			var footprintChars = footprintYaml != null ? footprintYaml.Value.Value.Where(x => !char.IsWhiteSpace(x)).ToArray() : new[] { 'x' };
+
+			var dimensionsYaml = yaml.Nodes.FirstOrDefault(n => n.Key == "Dimensions");
+			var dim = dimensionsYaml != null ? FieldLoader.GetValue<CVec>("Dimensions", dimensionsYaml.Value.Value) : new CVec(1, 1);
+
+			if (footprintChars.Length != dim.X * dim.Y)
+				throw new YamlException("Invalid footprint");
+
+			var index = 0;
+			var ret = new Dictionary<CVec, FootprintCellType>();
+			for (var y = 0; y < dim.Y; y++)
+			{
+				for (var x = 0; x < dim.X; x++)
+				{
+					var c = footprintChars[index++];
+					if (!Enum.IsDefined(typeof(FootprintCellType), (FootprintCellType)c))
+						throw new YamlException("Invalid footprint cell type '{0}'".F(c));
+
+					ret[new CVec(x, y)] = (FootprintCellType)c;
+				}
+			}
+
+			return ret;
+		}
+
+		public IEnumerable<CPos> FootprintTiles(CPos location, FootprintCellType type)
+		{
+			return Footprint.Where(kv => kv.Value == type).Select(kv => location + kv.Key);
+		}
+
+		public IEnumerable<CPos> Tiles(CPos location)
+		{
+			foreach (var t in FootprintTiles(location, FootprintCellType.OccupiedPassable))
+				yield return t;
+
+			foreach (var t in FootprintTiles(location, FootprintCellType.Blocking))
+				yield return t;
+
+			foreach (var t in FootprintTiles(location, FootprintCellType.BlockingUntargetable))
+				yield return t;
+		}
+
+		public IEnumerable<CPos> Tiles(Actor actor)
```

Comment:
This is only used from a handful of places, so it would be better to remove this and have them call `Tiles(CPos)` directly.